python/ray/experimental/channel/shared_memory_channel.py

```diff
@@ -518,15 +518,18 @@ def __init__(
 
         remote_reader_and_node_list: List[Tuple["ray.actor.ActorHandle", str]] = []
         for reader, node in self._reader_and_node_list:
-            if self._writer != reader:
+            if reader != self._writer:
                 remote_reader_and_node_list.append((reader, node))
         # There are some local readers which are the same worker process as the writer.
         # Create a local channel for the writer and the local readers.
         num_local_readers = len(self._reader_and_node_list) - len(
             remote_reader_and_node_list
         )
         if num_local_readers > 0:
-            local_channel = IntraProcessChannel(num_local_readers)
+            # Use num_readers = 1 when creating the local channel,
+            # because we have channel cache to support reading
+            # from the same channel multiple times.
+            local_channel = IntraProcessChannel(num_readers=1)
```

Comment:
Seems we can either remove IntraProcessChannel or make its read/write ops a no-op?

Reply:
My plan is to remove `IntraProcessChannel` and use `CachedChannel` instead.

Reply:
`CachedChannel` now should be able to replace `IntraProcessChannel` when the `inner_channel` is `None`.
However, I plan to do it in a follow up PR for a few reasons:
* The changes are modular: this PR focuses on introducing a caching mechanism and providing new capablity; while the next PR focuses on replacement/cleanup.
* Provide the caching feature earlier to users
* Replacing `IntraProcessChannel` needs quite some changes in code and tests. It is cleaner to separate the two.

Reply:
Hmm okay, sounds good. Btw, I think modifying IntraProcessChannel to have a no-op read might be a bit cleaner because then `CachedChannel` can always take a `ChannelInterface` instead of branching on whether one is provided or not. But I don't have a strong preference.

Reply:
Sounds like a good idea! I will probably take that approach. 

Will do it afterwards since it will be a backwards incompatible change and will break `IntraProcessChannel` tests.